src/commands/mod.rs

```diff
@@ -2411,6 +2438,95 @@ aborted.
     Ok(())
 }
 
+fn cmd_chmod(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    subcommand: &ChmodSubcommand,
+) -> Result<(), CommandError> {
+    let (executable_bit, executable_bit_str, args) = match subcommand {
+        ChmodSubcommand::Normal(args) => (false, "non-executable", args),
+        ChmodSubcommand::Executable(args) => (true, "executable", args),
+    };
+
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let repo_paths: Vec<_> = args
+        .paths
+        .iter()
+        .map(|path| workspace_command.parse_file_path(path))
+        .try_collect()?;
+    let commit = workspace_command.resolve_single_rev(&args.revision)?;
+    workspace_command.check_rewritable(&commit)?;
+
+    let tree = commit.tree();
+    let store = tree.store();
+    let mut tree_builder = store.tree_builder(tree.id().clone());
+    for repo_path in repo_paths {
+        let user_error_with_path = |msg: &str| {
+            user_error(format!(
+                "{}: {msg}",
+                workspace_command.format_file_path(&repo_path)
+            ))
+        };
+        let new_tree_value = match tree.path_value(&repo_path) {
+            None => return Err(user_error_with_path("can't find anything at this path")),
+            Some(TreeValue::File { id, executable: _ }) => TreeValue::File {
+                id,
+                executable: executable_bit,
+            },
+            Some(TreeValue::Conflict(id)) => {
+                let conflict = tree.store().read_conflict(&repo_path, &id)?;
+                let (new_removes, _) = chmod_conflict_sides(conflict.removes(), executable_bit);
+                let (new_adds, adds_include_files) =
+                    chmod_conflict_sides(conflict.adds(), executable_bit);
+                if !adds_include_files {
+                    return Err(user_error_with_path(
+                        "none of the sides of the conflict are files",
+                    ));
+                }
+                let new_conflict_id =
+                    store.write_conflict(&repo_path, &Conflict::new(new_removes, new_adds))?;
+                TreeValue::Conflict(new_conflict_id)
+            }
+            Some(_) => return Err(user_error_with_path("neither a file nor a conflict")),
+        };
```

Comment:
If we go this way, we'd probably want to check this per argument to the command, similarly to how we check revisions for being non-empty in `rebase`.

Reply:
Yes, ideally. Mercurial kind of does that, but only for prefix patterns (or something like that, I can't remember). For example, `hg st foo` will print an error if `foo` doesn't exist, but it's fine even if it's an empty directory [^1]. `hg st 'glob:**.foo'` is not an error, even if it doesn't match any patterns in the repo.

One possible way we could implement it is to pass in a list of matchers to `entries_matching()` and have it keep track of which matches have matched so far. Since the function returns an iterator, maybe the "which matchers have matched any paths" would need to be state that's updated in some struct passed in by the caller.

[^1]: Empty directories is a particularly tricky case to support in jj. I think we'll have to ignore that case since we don't track empty directories.